src/util/markdownToText.ts

```diff
@@ -0,0 +1,240 @@
+
+    let str = ''
+
+    for (const token of tokens) {
+      switch (token.type) {
+        case 'text':
+          str += token.text.replace(/\n/g, '&#10;')
+          break
+        case 'em':
+          str += `<i>${processInlineTokens(token.tokens ?? token.text)}</i>`
+          break
+        case 'strong':
+          str += `<b>${processInlineTokens(token.tokens ?? token.text)}</b>`
+          break
+        case 'del':
+          str += `<strike>${processInlineTokens(token.tokens ?? token.text)}</strike>`
+          break
+        case 'link':
+          if (token.href === token.text) str += token.text
+          else str += `<a href="${token.href}">${processInlineTokens(token.tokens ?? token.text)}</a>`
+          break
+        case 'image':
+          str += `<img src="${token.href}" alt="${token.text}" title="${token.title}" />`
+          break
+        case 'codespan':
+          str += `<code>${token.text}</code>`
+          break
+
+        default:
+          console.warn('Unknown inline token:', token.type)
+          break
+      }
+    }
+
+    return str
+  }
+
+  /** Processes markdown tokens recursively and incrementally builds the result. */
+  const processTokens = (tokens: Token[], parentDepth = 0) => {
+    // Stack of depths to determine indentation, scoped to the current parent
+    const stack: number[] = []
+
+    /** Helper function to get the indentation for a given depth. */
+    const indent = (depth: number) => '  '.repeat(depth + parentDepth)
+
+    for (const [index, token] of tokens.entries()) {
+      /** Helper function to check if ::/=scope is needed. */
+      const shouldScope = () => {
+        const prevHeadingIndex = tokens
+          .slice(0, index)
+          .reverse()
+          .findIndex(t => t.type === 'heading')
+        const prevHeadingActualIndex = prevHeadingIndex === -1 ? -1 : index - 1 - prevHeadingIndex
+
+        const nextHeadingIndex = tokens.slice(index + 1).findIndex(t => t.type === 'heading')
+        const nextHeadingActualIndex = nextHeadingIndex === -1 ? -1 : index + 1 + nextHeadingIndex
+
+        // Siblings are only considered at the same stack depth
+        const siblings = tokens.slice(
+          prevHeadingActualIndex + 1,
+          nextHeadingActualIndex === -1 ? undefined : nextHeadingActualIndex,
+        )
+
+        if (token.type === 'list') {
+          // If the list is ordered, we always need a scope if there are any other siblings or if we're at the top level.
+          if (token.ordered) return siblings.length > 1 || (parentDepth === 0 && prevHeadingActualIndex === -1)
+
+          // If the list is unordered, check if there are paragraph siblings
+          return siblings.some(t => t.type === 'paragraph')
+        } else if (token.type === 'table') {
+          // If the table is the only sibling or at the top level, we don't need a scope. Otherwise, we do.
+          return siblings.length > 1 || (parentDepth === 0 && prevHeadingActualIndex === -1)
```

Comment:
This appears to be the same condition that is used for `token.type === 'list' && token.ordered)`. See if you can combine them to avoid duplicating this logic.